qualtran/bloqs/gf_arithmetic/gf2_multiplication.py

```diff
@@ -653,3 +660,333 @@ def _binarypolynomialmultiplication() -> BinaryPolynomialMultiplication:
 _BINARY_POLYNOMIAL_MULTIPLICATION_DOC = BloqDocSpec(
     bloq_cls=BinaryPolynomialMultiplication, examples=(_binarypolynomialmultiplication,)
 )
+
+
+@attrs.frozen
+class GF2ShiftRightMod(Bloq):
+    r"""Multiplies by $2^k$ (or $x^k$ for polynomials) modulu.
+
+    Applies the transformation
+    $$
+        \ket{f} \rightarrow \ket{x^k f \mod m(x)}
+    $$
+
+    Where the modulus $m(x)$ is the irreducible polynomial defining the galois field arithmetic.
+
+    Args:
+        m_x: The irreducible polynomial that defines the galois field.
+        k: The number of shifts (i.e. the exponent of $2$ or $x$).
+
+    Registers:
+        f: The number (polynomial) to shift.
+
+    References:
+        [Space-efficient quantum multiplication of polynomials for binary finite fields with
+            sub-quadratic Toffoli gate count](https://arxiv.org/abs/1910.02849v2) Section 3.1
+    """
+
+    m_x: Poly = attrs.field(converter=lambda x: x if isinstance(x, Poly) else Poly.Degrees(x))
```

Comment:
Please consider taking the galois field type `GF` everywhere instead of taking `Poly` type. It makes the API simpler. You could also consider taking the `QGF` type directly, which encoded all the necessary information about the underlying field and represents the type of the quantum register on which you wish to perform the operations. 

The signature of all the bloqs would change to `dtype: QGF` and other parameters like `k: SymbolicInt` here and `const: const ` above for multiplication by a constant and `dtype: QGF` and `uncompute: bool` for `GF2Mul` bloq below.

Reply:
done